File: PO_py/program/Organisms/Animals/Turtle.py
import random
from program.Organisms.Animal import Animal



class Turtle(Animal):
    TURTLE_INITIATIVE = 1
    TURTLE_STRENGTH = 2

    def __init__(self, world, x, y):
        super().__init__(world, x, y, Turtle.TURTLE_INITIATIVE, Turtle.TURTLE_STRENGTH, 'T')
        print("Turtle created")

    def get_name(self):
        return "Turtle"

    def create_child(self, x, y):
        return Turtle(self.world, x, y)

    def action(self, command):
        if random.randint(0, 3) != 0:  # Turtle moves only 25% of the time
            return

        move = [0, 0]
        if not self.randomize_position(move):
            return

        new_x = self.x + move[0]
        new_y = self.y + move[1]

        if not self.world.is_position_empty(new_x, new_y):
            if self.collision(self.world.get_organism_at_position(new_x, new_y)):
                self.move(move[0], move[1])
        else:
            self.move(move[0], move[1])

    def dodge(self, attacker, attacked):
        if attacker.get_strength() < 5:
            print("Turtle defended itself from the attack")
            return True
        return False

File: PO_java/program/Organisms/Animals/Turtle.java
package projekt.program.Organisms.Animals;

import projekt.program.Organism;
import projekt.program.Organisms.Animal;
import projekt.program.World;

import java.util.Random;

public class Turtle extends Animal {

    private static final int TURTLE_INITIATIVE = 1; // Example value
    private static final int TURTLE_STRENGTH = 2; // Example value

    public Turtle(World world, int x, int y) {
        super(world, x, y, TURTLE_INITIATIVE, TURTLE_STRENGTH, 'T');
        System.out.println("Turtle created");
    }

    @Override
    public String getName() {
        return "Turtle";
    }

    @Override
    protected Organism createChild(int x, int y) {
        return new Turtle(this.world, x, y);
    }

    @Override
    public void action(char command) {
        Random random = new Random();
        if (random.nextInt(4) != 0) { // Turtle moves only 25% of the time
            return;
        }
        int[] move = new int[2];
        if (!randomizePosition(move)) {
            return;
        }
        int newX = this.x + move[0];
        int newY = this.y + move[1];
        if (!world.isPositionEmpty(newX, newY)) {
            if (collision(world.getOrganismAtPosition(newX, newY))) {
                move(move[0], move[1]);
            }
        } else {
            move(move[0], move[1]);
        }
    }

    @Override
    public boolean dodge(Organism attacker, Organism attacked) {
        if (attacker.getStrength() < 5) {
            System.out.println("Turtle defended itself from the attack");
            return true;
        }
        return false;
    }
}

